### PR TITLE
adds the possibility to create separatly Query object to use it in parallel in different threads

### PR DESCRIPTION
**Pull request by dpankrat**
I simply added a new abstarct class LSHNearestNeighborQuery like LSHNearestNeighborTable which wraps Query. I couldn't add a metthod to LSHNearestNeighborTable which returns directley Query object, because Query couldn't be resolved in lsh_nn_table.h (we need forward declaration perhaps?).

### CODE REVIEW

**Comment by ludwigschmidt** [comment]
Thanks a lot for the pull request! Overall this is a great addition. Two points:
- You gave the new externally-visible query object the name `LSHNearestNeighborQuery`. So far, the code is structured so that "NearestNeighbor" query objects also have functions based on computing actual distances (not only producing LSH candidates). Since the new `LSHNearestNeighborQuery` only wraps a `Query` object internally, it would make more sense to call it `LSHQuery`. An alternative (maybe even better) would be to wrap a NearestNeighborQuery object and add the corresponding functions to the new `LSHNearestNeighborQuery`.
- We have roughly followed the Google style guide for the C++ code so far. The new code doesn't match the indentation of the current code. Could you please fix that? Thanks and sorry for being picky about this! We hope in the long term it helps to make the code more maintainable.

**Comment by ilyaraz** [comment]
For formatting the code, one can just run `clang-format --style=Google -i <file-name>`.

**Comment by ilyaraz** [comment]
Can we call the function `construct_query` differently? Maybe `construct_query_processor` or something similar? Right now the name is quite misleading.

**Comment by dpankrat** [comment]
Yes, you are right, now it is ambiguous. I will rename it.

**Comment by dpankrat** [comment]
Hi! I made these changes and commited it. (I made a wrapper of NearestNeighborQuery, I think it is really better.) Cheers!Dima 26.10.2016, 01:41, "Ilya Razenshteyn" notifications@github.com:For formatting the code, you can simply run clang-format --style=Google -i <file-name>.—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by ludwigschmidt** [comment]
Looks great! I just fixed a small typo.

Just for consistency: can we name it `construct_query_object`? I think in the rest of the code we use the term query object instead of query processor.

**Comment by ludwigschmidt** [comment]
@ilyaraz We have to make sure we're not breaking the Python interface with this change. The Python wrapper wraps the C++ wrapper this pull request is changing.

Are there cases where query objects would make sense in Python? Since Python is single-threaded due to the GIL, I don't see a strong use case (but I might just be missing it).

If per-thread query objects don't make sense in Python land, we might want to tell swig to ignore that part of the C++ wrapper.

**Comment by dpankrat** [comment]
Of course you can))

**Comment by ilyaraz** [comment]
How does the script from `ann-benchmarks` issues queries in threads then? I guess wrapped C++ code can be executed in parallel, but the wrapper code can not. So, including the query objects to the Python wrapper might make sense. But let's do it in a separate commit, and tell swig not to do anything for now.

**Comment by ludwigschmidt** [comment]
Yes, the C++ code called from Python can use multiple threads (and I guess that's how the other libraries in ann-bechmarks deal with this). But the current code doesn't actually use multiple threads on the C++ side, so using the new query object from Python wouldn't give you any parallelism.

I think there are three changes remaining:
1. Rename to `query_object` instead of `query_processor`.
2. Add a test to `src/test/cpp_wrapper_test.cc` (ideally multi-threaded).
3. Tell swig not to wrap the new interface.

I can take care of point 3. Point 1 should be easy. Any volunteers for point 2? :-)

**Comment by dpankrat** [comment]
I can do the point 2, but only after serialization. What are the rules for writing such tests? I have never written it before.
I don't think that we really need multi-threaded test, _const_ reference to data (without _mutable_ members) will guaranty it.

**Comment by ilyaraz** [comment]
That's how ann-benchmarks do it.

``` python
def single_query(t):
    v, correct = t
    found = algo.query(v, 10)
    return len(set(found).intersection(correct))
if algo.use_threads():
    pool = multiprocessing.pool.ThreadPool()
    results = pool.map(single_query, queries)
else:
    results = map(single_query, queries)
```

More context for this: http://stackoverflow.com/questions/3033952/python-thread-pool-similar-to-the-multiprocessing-pool . So there seem to be proper threads, which allow to execute at least C++ code in parallel.

**Comment by ludwigschmidt** [comment]
OK, interesting. Do we have to release the GIL ourselves in the wrapper then?

**Comment by ilyaraz** [comment]
Apparently not. I doubt that the current libraries do it. Probably when CPython passes the control to an extension, it releases GIL itself, but obviously I really have no idea.

**Comment by ludwigschmidt** [comment]
@dpankrat There are no particular rules regarding our tests. Just a simple test that creates a query object, uses it for a lookup, and then checks the answer should be fine. Since FALCONN uses templates heavily, it's good if the tests check that everything compiles at least :-)

A multi-threaded test would be nice (and wouldn't be too hard with the C++11 threading classes). But I agree that things should be fine in this case.

**Comment by ludwigschmidt** [comment]
@ilyaraz At least annoy wraps things in `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS` on the C++ side, so it explicitly releases the GIL.

It looks like SWIG also has functionality for releasing the GIL.

**Comment by ludwigschmidt** [comment]
Thanks a lot for the new changes! I'll look into the SWIG wrapper this weekend. Then we can merge this :-)

**Comment by ilyaraz** [comment]
So, any progress on this front? When the SWIG wrapper is done, I can take care of the "outer" Python wrapper.

**Comment by ludwigschmidt** [comment]
Sorry for being slow on this. I've been busy with application stuff unfortunately and this week there is a lot happening in 6.006. But I hope I'll find some time.

**Comment by ilyaraz** [comment]
I'm not sure why I "closed" this, but it obviously was a misclick. Reopening...

File: src/include/falconn/lsh_nn_table.h [hunk 1]
```diff
@@ -88,36 +79,158 @@ class LSHNearestNeighborTable {
...
+  
+  ///
+  /// This function constructs a new query pool. The query pool holds
+  /// a set of query objects and supports an interface that can be safely
+  /// called from multiple threads.
+  ///
+  virtual std::unique_ptr<LSHNearestNeighborQueryPool<PointType, KeyType>>
+  construct_query_pool(int_fast64_t num_probes = -1,
+                       int_fast64_t max_num_candidates = -1,
+                       int_fast64_t num_query_objects = -1) const = 0;
```
Comment (ilyaraz) [inline src/include/falconn/lsh_nn_table.h]: What's the semantics of `num_query_objects` being equal to -1?

Comment (ludwigschmidt) [reply src/include/falconn/lsh_nn_table.h]: It uses a small constant factor times the number of available hardware threads. This should be sufficient for use in a good thread pool implementation (that has number of parallel threads = number of hardware threads).

Comment (ilyaraz) [reply src/include/falconn/lsh_nn_table.h]: I'd say it's worth being documented. What does the constant equal to btw?

Comment (ludwigschmidt) [reply src/include/falconn/lsh_nn_table.h]: The constant is 2. I agree that we should document this.

**Comment by chapleau** [comment]
Thanks for working on such a useful features. Is there an ETA on this ? The ability to query safely  in parallel (from python) would be awesome !

**Comment by ludwigschmidt** [comment]
We have a working prototype in C++ now. But unfortunately we haven't had time to write the Python wrapper so far. Ilya and I are quite busy at the moment. Hopefully we'll get to it in a month or so.

**Comment by ilyaraz** [comment]
I've just run the random benchmark with 1--4 cores on my machine (with i7-6700K). Here are the speed-ups we are getting:
*  1 core -- 1.000000
* 2 cores -- 1.968170
* 3 cores -- 2.823859
* 4 cores -- 3.732441

Are we happy about it? Can we push it closer to 4.00?

**Comment by ilyaraz** [comment]
For GloVe, the situation is much worse:
* 1 core -- 1
* 2 cores -- 1.82775
* 3 cores -- 2.48087
* 4 cores -- 2.96792

**Comment by ludwigschmidt** [comment]
Interesting, I also got close to 4x on the random benchmark. I haven't tried GloVe yet. Two questions:

- Is this the C++ or Python interface?

- How does the "time for distance computation" vs "time for everything else" balance look for the two tests you ran? Generally we spend more time in distance computation on GloVe. Are we hitting some memory bottleneck?

**Comment by ilyaraz** [comment]
* This is the C++ interface. Before working on the wrapper, it's good to know how well the C++ code scales.
* It seems that the distance computation takes around 70% of the overall time. I first thought that it might be the case that the total number of candidates is quite different, but it does not seem to be the case. So looks like, the distance computation is a bottleneck of some sort?

**Comment by ilyaraz** [comment]
Yes, just computing the list of candidates scales a good deal better:
* 1
* 1.95269
* 2.82279
* 3.62436

And it seems like the heap is OK (for finding k smallest distance), so the problem is really computing distances with candidates.

I wonder if prefetching becomes less efficient??

**Comment by ludwigschmidt** [comment]
I agree that we should understand the performance of the C++ code first.

It might simply be the case that we're exhausting the memory bandwidth available to the 4 cores. Do you know what the total memory bandwidth is and what we're currently using?

Also, going to more than 4 threads might bring us closer to the 4x hardware limit. I assume your CPU has hyper-threading, so it should be able to hide some of the memory latency.

**Comment by ilyaraz** [comment]
Those are good questions, I'll try to answer them later today.

**Comment by ilyaraz** [comment]
Interesting, there is something I can't make sense of. Per thread we have around 45M candidates, which we process in under a second, which means that the total memory bandwidth (for all the cores) must be at least 70 Gb/s. However, [this page](https://ark.intel.com/products/88195/Intel-Core-i7-6700K-Processor-8M-Cache-up-to-4_20-GHz) says that the maximum memory bandwidth is 34.1 Gb/s. What's going on? Is the cache that efficient? Hard to believe.

UPD: nevermind, it takes 9 seconds, not 0.9. So, we exhaust only 7 Gb/s out of 34.1 Gb/s of the bandwith.

**Comment by ludwigschmidt** [comment]
OK, interesting. I guess the bandwidth is for an ideal scenario though (linear scan with lots of pre-fetching).

Overall, does more of the time shift to distance computation as you increase the number of threads? You mentioned 70% of the overall time is spent on distance computation. I guess for 1 thread that's less then?

**Comment by ilyaraz** [comment]
I've written a [micro-benchmark](http://pastebin.com/hP4y4mxu) that emulates distance computations from several threads, when we access data points in random order.

We obtain the following speed-ups (with hyperthreading turned off in BIOS):
* 1
* 1.46999
* 1.76156
* 1.95018

Alas, when turned on, hyperthreading does not seem to help and, moreover, it makes the running times with <= 4 threads somewhat worse.

Overall, we achieve the bandwidth of 9.8 Gb/s, when we use 4 threads, which is comparable to the above results. I guess it means that, whenever we tune the parameters, we should discourage having more candidates, as long as we query tables from several threads.

**Comment by ilyaraz** [comment]
Thinking a bit more about it, we could seriously benefit from 16- or 8-bit floats. It should improve the distance computation step by a bit, uncongesting the memory bottleneck. What do you think?

**Comment by ludwigschmidt** [comment]
Cool, thanks for looking into this. I wonder if there are better ways of prefetching since we know the exact list of candidates.

Would it be possible to repeat your micro-benchmark, this time with candidate sets that are intervals? Then every thread would do look-ups in order and we could see whether better pre-fetching could help.

**Comment by ilyaraz** [comment]
Sure, after I'll finish my ICALP reviews, which are seriously overdue. :)

**Comment by ludwigschmidt** [comment]
Lower accuracy floats are an interesting idea. However 16 bit floats don't seem to be part of standard C++ (numpy has a float16 though). We could also try integers with few bits?

**Comment by ludwigschmidt** [comment]
Oops, thanks for reminding me about the ICALP reviews :-)

**Comment by ilyaraz** [comment]
Here are the results, when we sort segments of a given length: 1 (that is, don't sort), 10k, 20k, and 40k. The numbers are running times. We prefetch two rows ahead, as before.

| Sorted segments | 1 thread | 2 threads | 3 threads | 4 threads|
|---|---|---|---|---|
|1|7.38234|5.03459|4.2082|3.81201|
|10000|6.01061|4.29058|3.68817|3.51019|
|20000|5.84528|4.2165|3.57508|3.36628|
|40000|5.56234|4.1416|3.47031|3.24105|

As for the float16, I've just tried it in numpy, and, unless I've done something terribly wrong, it's super slow. So, the point seems to be saving memory, not improving speed.

**Comment by ludwigschmidt** [comment]
Cool, thanks for the numbers. Just to avoid confusion: are the candidates now **contiguous** and sorted subsets of length 10k etc.?

I don't understand why float16 is much slower than float32. In any case, we should be able to use int16 per coordinate for normalized vectors in moderate dimension (say 100 - 1000)? We can discretize each coordinate to the closest multiple of +/- 1/2^15.

**Comment by ilyaraz** [comment]
> are the candidates now contiguous and sorted subsets of length 10k etc.?

Yes.

> we should be able to use int16 per coordinate for normalized vectors in moderate dimension (say 100 - 1000)? We can discretize each coordinate to the closest multiple of +/- 1/2^15.

Sure, but, for example, Eigen does not vectorize operations with 16-bit integers, so we'd need to implement it manually.

**Comment by ludwigschmidt** [comment]
Hmm, to what extent does Eigen not vectorize 16-bit integers? Part of their website claims they do:

http://eigen.tuxfamily.org/dox/TopicCustomizing_CustomScalar.html

> By default, Eigen currently supports standard floating-point types (float, double, std::complex<float>, std::complex<double>, long double), as well as all native integer types (e.g., int, unsigned int, short, etc.), and bool.

AVX2 also has the right instructions:

https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=AVX2&cats=Arithmetic&expand=64

Of course, we need to check what actually happens. There is also the issue that we might want the arithmetic to happen with 32bit integers so we don't get too bad overflows?

**Comment by ilyaraz** [comment]
In the end, figured out that my CPU does not support RAM frequency more than 2133 MHz, which gives bandwidth of 15.9 Gb/s, the micro-benchmark achieves bandwidth 11.5 Gb/s. I wonder if we can close the gap.

**Comment by ludwigschmidt** [comment]
I guess the most memory friendly access pattern would be a linear scan over an array of integers (or another simple type). The current micro-benchmark has one level of indirection in the memory organization (we have a vector of Point which contain a pointer each). Could you try running the micro-benchmark with a single large array per worker (no indirections)? Then we'd know how much the indirections are hurting us, and what the simple linear scan baseline is.

**Comment by ilyaraz** [comment]
If we make each chunk of candidates of length T *consecutive* numbers, then starting from T around 500, we pretty much saturate the bandwidth (even exceed it a little bit, probably because of some cache-related effects).

I've also asked one expert in multi-core implementations, and he told me that not saturating the full bandwidth when accessing memory at random is OK, and the utilization we are getting is actually pretty good.

**Comment by ludwigschmidt** [comment]
OK, that's good to know. Thanks!

I guess one explanation is that the memory system is more efficient for accessing large blocks of data. One 128-dim vector is only 512 bytes.

Do things improve when we have all points in a dense matrix (as opposed to a vector<Point>)? If it makes a noticeable difference, we might want to encourage contiguously stored data in the C++ interface.

**Comment by ilyaraz** [comment]
OK, now to Python wrapper. If I understand correctly, we now have a SWIG wrapper updated, but not the higher-level wrapper. What's a new interface gonna look like?

I think it make sense to have two mode. The first as before: we create a table, and just use it. The second should be, when the underlying C++ code spawns several threads, and we feed queries to it. The latter should be much more efficient than using Python threads.

**Comment by ludwigschmidt** [comment]
The C++ code currently supports two ways to issue queries:

1. Get a query object from the LSHTable. That query object is valid for use in a single thread.

2. Get a query pool from the LSHTable. The query pool can be used from multiple threads (internally it has a set of query vectors that it uses for different threads).

I wrote interface 2 so we can support threads on the Python side (as is done in ann-benchmarks). One approach for supporting the current high-level Python interface could be the following: when we create the LSHTable object, we also create a query pool. All of the current ANN calls are then forwarded to the internal query pool. That way our high-level Python interface is always thread safe. We'll have to set a reasonable default for the number of threads, but that should be possible.

I like your idea of a third interface. Basically you're proposing a "batch query" interface where we are doing the multithreading internally, right?

If you get a chance, it might still be good to know what happens with multi-threaded performance with big-matrix vs vector<Point> storage.

**Comment by ludwigschmidt** [comment]
I added #74 so we don't forget the 8 or 16 bit integer quantization idea. It might be low-hanging fruit performance-wise.

**Comment by ilyaraz** [comment]
BTW, I educated myself a bit, and realized that the reason we don't get full bandwidth is likely to be [TLB](https://en.wikipedia.org/wiki/Translation_lookaside_buffer) misses. My CPU has TLB consisting of 64 entries for 4K pages, which amounts to around 500 100-dimensional vectors, like what I noticed above.

**Comment by ludwigschmidt** [comment]
OK, interesting. So in order to achieve the memory bandwidth, we'll have to access *all* of the page that's referenced in the TLB? So if the number of accesses per page is too small, the address translation becomes the bottleneck.

**Comment by ilyaraz** [comment]
Yes, alternatively, we can use large pages. There is even a possibility of using up to 1 Gb pages via [HugeTLBFS](https://lwn.net/Articles/375096/), that is, we can fit a dataset into a single page!

[This](https://www.pvk.ca/Blog/2014/02/18/how-bad-can-1gb-pages-be/) is a relevant link.

**Comment by ludwigschmidt** [comment]
1 GB pages look interesting. The second article however says

> With multiple processes in play, there are fragmentation issues, and things aren’t as clear-cut… especially given that 1GB pages must currently be allocated at boot-time, on Linux.

Is that still accurate? It would be good if we could enable 1 GB pages for our process only.

**Comment by ilyaraz** [comment]
Yes, in Linux there is a separate notion of "huge pages", you can configure their size and amount, and they are allocated during the boot. Then, one can manually allocate these pages, if I understood correctly.

**Comment by ludwigschmidt** [comment]
Hmm, then I guess it's not ideal because a simple "pip install" won't be able to make these changes for you, correct?

**Comment by ilyaraz** [comment]
Sure, but we can provide a flag so that we can start using huge pages, if a user enabled them. It requires a bit of expertise, though.

**Comment by ilyaraz** [comment]
This does not make any sense. I re-implemented the above micro-benchmark to simplify the further experiments. The differences are:
* Instead of using a vector of Eigen's vector, I use a contiguous array of floats;
* Instead of using Eigen's dot product, I've implemented my own using AVX;
* All the points are 32-byte aligned (in particular, that's why I use dimension 104 instead of 100).

And even if we use a single thread with random candidates and without prefetching, it all takes only 4.77 seconds instead of 7.38. I wonder what causes such a difference.

The new version is [here](https://pastebin.com/RLPqrPws).

**Comment by ilyaraz** [comment]
With 4 threads it takes 2.51 seconds, which means that we use the bandwidth for about 97%! So, something is clearly wrong with using std::vector of Eigen's vectors...

P.S. A multi-threaded version of the benchmark is [here](https://pastebin.com/D8u4KmY2).

**Comment by ludwigschmidt** [comment]
Very interesting! Do we know which of the three changes has a performance effect?

- What happens if you replace the custom dot-product with an Eigen Map<VectorXf> (see http://eigen.tuxfamily.org/dox/group__TutorialMapClass.html )?

- What happens if you use dimension 100 instead of 104?

If Eigen's dot product via Map<VectorXf> is slower than handwritten AVX, we should file a bug report with Eigen.

**Comment by ilyaraz** [comment]
If I wrap the array into Map<MatrixXf>, then I don't understand how to perform dot products between rows without much overhead: the naive code a.row(i).dot(b.row(j)) takes around 70 seconds! If we wrap it into vector<Map < VectorXf> >, then it's much faster, but overall, a bit slower than the initial benchmark (so the manual AVX is much faster). So, Eigen is being problematic. :(

I'm trying to debug it to understand what causes the overhead. But if it confirms, we might want to switch to using OpenBLAS (dot product is one of the BLAS level 1 operations).

**Comment by ludwigschmidt** [comment]
Eigen stores matrices in column-major order by default (see https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html ). So doing a dot product with a row of the matrix has very bad locality. I guess this explains the slow running time.

What happens if you don't build a vector<Map<VectorXf>> but instead create the Map<VectorXf> when you use it? That way we we don't have the vector accesses. And ideally the Map<VectorXf> object should be free in terms of running time (completely optimized away).

I agree that we should look into OpenBLAS for the distance computation at least if Eigen turns out to be problematic. But it might be good to understand what Eigen produces first. Did you have a chance to look at the assembly code? See http://eigen.tuxfamily.org/index.php?title=Developer%27s_Corner#Studying_assembly_output

Also, can you include the microbenchmark in the repo? It might be good to have a microbenchmark folder with simple baselines for the main lookup parts (hasing, table lookup, distance computation). That way we can make sure in the future that we don't get unnecessary overhead.

**Comment by ilyaraz** [comment]
That indeed does it (if we wrap into Map<VectorXf> on the fly)! Here is the comparison:

| Method | 1 thread | 2 threads | 3 threads | 4 threads|
|---|---|---|---|---|
|Hand-written AVX (104 dimensions) |4.97919|3.439|2.843|2.50684|
|Eigen (104 dimensions)|5.05227|3.37071|2.77908|2.45958|
|Eigen (100 dimensions)|5.44095|3.59269|2.95201|2.6035|

Takeaways:

* For one thread hand-written AVX is more efficient, but for four threads, Eigen utilizes the memory bandwidth pretty much perfectly;
* For 100 dimensions, things are noticeably slower than for 104, funnily enough (but unsurprisingly).

**Comment by ludwigschmidt** [comment]
OK, great that Eigen works well in that configuration. The downside is that we don't get an unexpected speed boost :-)

For the plain array data storage, we're already using a Map that is created on the fly:

https://github.com/FALCONN-LIB/FALCONN/blob/master/src/include/falconn/core/data_storage.h#L222

It would still be good to have the microbenchmark in the repo I think. Then we can compare the microbenchmark numbers (number of inner products per second) with the numbers reported in FALCONN.

**Comment by ilyaraz** [comment]
I did add a benchmark yesterday. Also, we can get a tiny bit of an improvement by allocating an aligned chunk of memory and by padding the number of dimensions to a multiple of 8. :) Also, if we will be running on 8 threads, we'll need to re-tune parameters a little bit, which should also help. OK, I'll start working on a Python wrapper then.

**Comment by ilyaraz** [comment]
A comparison of `vector<VectorXf>` vs `float*` for GloVe:

| Method | 1 thread (ms) | 2 threads (ms) | 3 threads (ms) | 4 threads (ms) |
|---|---|---|---|---|
|`vector<VectorXf>` |2.86|1.56|1.15|0.97|
|`float*`|2.33|1.29|0.97|0.82|

**Comment by ludwigschmidt** [comment]
Good to know, thanks! So not a huge difference but definitely noticeable (in the AVX vs Eigen benchmarks I wasn't sure if it's within statistical error).

**Comment by ilyaraz** [comment]
Further profiling on GloVe gives quite interesting [results](https://pastebin.com/XBteF5b5).

In particular, I don't understand why "hash table time" is so high: it is comparable to the distance computation time!

**Comment by ilyaraz** [comment]
Interesting. Turns out the bit packed table is quite slow. [Here](https://pastebin.com/Y8M4USCs) are the results without bit packing. For 4 threads, the bit packed version is roughly twice as slow.

I still don't fully understand why the hash table step is so slow. Could it be because of the `push_back` we are doing (potential memory allocations are not very multi-thread-friendly)?

**Comment by chapleau** [comment]
Will this approach scale well with the number of threads used in the case where there are a lot of candidates to sort ? The distance computation is probably taking most of the time so I guess I'm asking if the distance computation is also multithreaded. 
Actually, does 'get_unique_candidates' sort by distances at all ? If in my current application, the actual distance computation is (I think?) the most time consuming, could one use get_unique_candidates and sort (in parallel for a batch of queries) outside Falconn ?
Thanks !

**Comment by ludwigschmidt** [comment]
Currently the multi-threaded implementation processes a query in a single thread (allowing multiple queries to be processed in parallel). You are right that ideally we'd also have an option to multi-thread a single query. I hope we can add this in the future.

get_unique_candidates does indeed not compute distances. So you could multi-thread the distance computation outside FALCONN yourself.

**Comment by chapleau** [comment]
Thank you for the prompt response. I agree that there are most likely more uses cases for multiple queries to be run in parallel, as oppose to a single one. Looking forward to trying that !

**Comment by ilyaraz** [comment]
I've just fixed the Python wrapper to make it work with the new C++ code. I left it thread-unsafe for now (the preprocessing might use many threads, but one can't query the same data structure from many threads), since we plan to upgrade the Python interface substantially anyways.

**Comment by ludwigschmidt** [comment]
🎉 👍

**Comment by chapleau** [comment]
Is this new feature planned to be released in v1.3 ? 
Thanks !

**Comment by ilyaraz** [comment]
It is already in the master branch. We updated the examples and documentation to accommodate that change. If you have any further questions, let us know.

**Comment by chapleau** [comment]
Thanks for the prompt answer. What about your last comment where it says 
that the python wrapper is still thread-unsafe, is that still the case 
in the master branch ?

**Comment by ilyaraz** [comment]
It is thread-safe now, but instead of `pip install falconn`, you should clone the repo and run `make python_package_install`.

**Comment by chapleau** [comment]
sounds good, thank you.

**Pull request merged**